lib/api.js

```diff
@@ -0,0 +1,145 @@
+const u = require('./util')
+const EventEmitter = require('events')
+// @ts-ignore
+const Hookable = require('hoox')
+
+/**
+ * @template T
+ * @param {T} x
+ * @returns {T}
+ */
+function identity (x) {
+  return x
+}
+
+/**
+ * @param {any} a
+ * @param {any} b
+ * @param {any=} mapper
+ */
+function merge (a, b, mapper) {
+  mapper = mapper ?? identity
+  for (const k in b) {
+    if (
+      b[k] &&
+      typeof b[k] === 'object' &&
+      !Buffer.isBuffer(b[k]) &&
+      !(b[k] instanceof Uint8Array) &&
+      !Array.isArray(b[k])
+    ) {
+      a[k] = {}
+      merge(a[k], b[k], mapper)
+    } else {
+      a[k] = mapper(b[k], k)
+    }
+  }
+  return a
+}
+
+/**
+ * @param {any} plugins
```

Comment:
is this an Array ? can you set that with a `@param`

Reply:
Yes. But note that there is no limit to how much more specific these types can be. I'll make it an `Array<any>`, but it could well be an `Array<Plugin>` where we specify what a Plugin is.

Actually, that's what it should ideally be, and I agree we should go towards more types (or more documentation), but it doesn't have to all be done within this PR because that takes a hell ton of time.

Reply:
In this case, done.